### PR TITLE
Interesting project - wanted to contribute

### PR DESCRIPTION
**Pull request by timoguic**
Possible improvements for review.

I originally wanted to add YAML parsing for frontmatter (I am used to it in Hugo & Jekyll). Is it something that is worth considering, or are you fixed on JSON?

### CODE REVIEW

File: snek/snek.py [hunk 1]
```diff
@@ -251,20 +251,23 @@ def __load_sitemap(self):
...
 
-            # Load from front-matter and add to default
-            try:
-                metadata_from_file = frontmatter.load(filepath, handler=frontmatter_json_handler()).metadata
-                for key, value in metadata_from_file.items():
-                    metadata[key] = value
-
             # If the file could not be read or open, go to next file
-            except FileNotFoundError:
+            if os.access(filepath, os.R_OK):
```
Comment (matteocargnelutti) [inline snek/snek.py]: Good catch 👍 !

File: snek/snek.py [hunk 2]
```diff
@@ -251,20 +251,23 @@ def __load_sitemap(self):
...
+
+            # Load from front-matter and add to default
+            try:
+                metadata_from_file = frontmatter.load(filepath, handler=frontmatter_json_handler()).metadata
             # If the file's content is not valid JSON frontmatter, simply log it
             except json.decoder.JSONDecodeError as err:
                 self.__add_error(f"{filepath} does not contain valid JSON. {err}")
 
+            # We don't want `filepath` from metadata to change our dict
+            metadata.pop("filepath", None)
```
Comment (matteocargnelutti) [inline snek/snek.py]: Good catch here too 👍 . Having `filepath` in the front-matter could have messed the process up.

File: snek/snek.py [hunk 3]
```diff
@@ -251,20 +251,23 @@ def __load_sitemap(self):
...
+            
+            # Update metadata with metadata from file
+            metadata.update(metadata_from_file)
```
Comment (matteocargnelutti) [inline snek/snek.py]: This is what happens when you spend too much time working with JavaScript 😆 . Much more Pythonic 🐍 .

File: snek/snek.py (hunk 1 above)
Comment (timoguic) [reply snek/snek.py]: Actually I made a mistake (just committed). Should be `if not`...

Comment (matteocargnelutti) [reply snek/snek.py]: This would likely need to be `if not os.access(filepath, os.R_OK)` though.

**Comment by matteocargnelutti** [comment]
@timoguic First of all, thank you very much for this PR.

I am very open to the idea of supporting YAML, especially since we already have everything we need to implement this feature.

I guess the idea would be to support YAML for both content files _(which could then be either `.json.md` or `.yml.md`)_ and shared data files _(`.json` or `.json.md`)_ and loosely detect which format is used in [`snek.__load_data()`](https://github.com/matteocargnelutti/snek/blob/master/snek/snek.py#L133), [`snek.__load_sitemap()`](https://github.com/matteocargnelutti/snek/blob/master/snek/snek.py#L220) and [`snek.__build_content`](https://github.com/matteocargnelutti/snek/blob/master/snek/snek.py#L353) ?

Would you like to undertake the development of this feature?
Happy to discuss / brainstorm ideas as needed.

**PS:** Ça fait plaisir :).

**Comment by timoguic** [comment]
Sounds good. So far I just made a quick hack/workaround to grab YAML frontmatter data instead of JSON if it is available. Do you want me to work on another PR or keep using this one?

**PS:** Mais oui! :)

**Comment by matteocargnelutti** [comment]
@timoguic Let's continue on this PR :) ! Thank you very much, let me know if you need anything.

**Comment by timoguic** [comment]
I made some commits, feel free to let me know what you think. Not sure about the use of `__` - is it your convention for private methods in this project?

**Comment by matteocargnelutti** [comment]
Thank you very much for this: I will have a closer look as soon as possible.

Just wanted to give you some context on the `__` convention for private class members: It is the  [recommendation from Python's documentation](https://docs.python.org/3/tutorial/classes.html#private-variables). Of course - as you know - there is no mechanism to enforce encapsulation in Python, this is just a naming convention.

**Comment by timoguic** [comment]
I think this part of the documentation is a bit unclear - PEP8 is much clearer (https://www.python.org/dev/peps/pep-0008/#id47):

```
Use one leading underscore only for non-public methods and instance variables.

To avoid name clashes with subclasses, use two leading underscores to invoke Python's name mangling rules.
```

More specifically: `Generally, double leading underscores should be used only to avoid name conflicts with attributes in classes designed to be subclassed.`

I don't think that is the use case here, but I don't want to be splitting hairs on conventions :)

**Comment by matteocargnelutti** [comment]
@timoguic Good point :). I guess it never bothered me as this is not caught by PyLint.

I really don't mind switching to single underscores if you'd like, it actually makes sense.

File: snek/snekconfig.py [hunk 4]
```diff
@@ -131,6 +136,19 @@ def __init__(self,
         self.scss_active = True if scss_active == True else False
         self.scss_output_style = scss_output_style if scss_output_style in ['compressed', 'nested', 'expanded'] else 'compressed'
         self.data_in_build = True if data_in_build == True else False
+        self.handlers = {
```
Comment (matteocargnelutti) [inline snek/snekconfig.py:139]: Really like this approach

File: snek/snekconfig.py [hunk 5]
```diff
@@ -131,6 +136,19 @@ def __init__(self,
...
+                'frontmatter_handler': JSONHandler,
+                'exception': json.decoder.JSONDecodeError,
+            },
+            '.yaml': {
+                'loader': yaml.safe_load,
+                'frontmatter_handler': YAMLHandler,
+                'exception': yaml.YAMLError,
+            }
+        }
+        self.handlers['.yml'] = self.handlers['.yaml']
```
Comment (matteocargnelutti) [inline snek/snekconfig.py:153]: This line could use a comment to explain its purpose

File: snek/snek.py [hunk 6]
```diff
@@ -130,6 +129,235 @@ def __add_error(self, message):
...
+        -------
+        list
+        """
+
+        glob_pattern = "**/*"
+
+        if extra_suffix and extra_suffix.startswith('.'):
+            glob_pattern = f"{glob_pattern}{extra_suffix}"
+
+        return [
```
Comment (matteocargnelutti) [inline snek/snek.py]: I would maybe not use a list comprehension for this, for readability purposes.
I kind of feel that as soon as you need to spread a list comprehension on multiple lines, it somewhat defeats its purpose.

File: snek/snek.py [hunk 7]
```diff
@@ -130,6 +129,235 @@ def __add_error(self, message):
...
+        if not isinstance(fm_data.handler, allowed_handlers):
+            self.__add_error(f"{filepath} has invalid frontmatter data.")
+            return None
+
+        if metadata_only:
+            return fm_data.metadata
+        else:
+            return fm_data
+
+    def __parse_data_from_filepath(self, filepath):
```
Comment (matteocargnelutti) [inline snek/snek.py]: Since this method is only used in `__update_data_from_filepath()`, I would merge the two.

File: snek/snek.py [hunk 8]
```diff
@@ -369,61 +506,66 @@ def __build_content(self):
...
+                                    content=page.content)
 
-                # Count as built
-                self.pages_built += 1
+            # Write HTML file
+            destination_dirname = os.path.dirname(destination_filepath)
+            if not os.path.exists(destination_dirname):
+                os.makedirs(destination_dirname)
+
+            with open(destination_filepath, 'w') as fp:
```
Comment (matteocargnelutti) [inline snek/snek.py:451]: Is there a technical advantage of using `with` in that context over `open(destination_filepath, 'w').write(html)`, given that in the latter the file pointer will also be removed as soon as the operation is done?

**Comment by matteocargnelutti** [comment]
@timoguic Just had a look: Thank you very much for all this 👍 .

Having YAML support is great and most of the edits you worked on are clear improvements.

Part of the idea of `snek` is to remain ["simple as possible, end to end"](https://github.com/matteocargnelutti/snek#snek-is-a-static-site-generator--framework-that-was-designed-with-two-ideas-in-mind): As such, there might be things I should probably have indicated were simplified on purpose, and which I am more than happy to discuss / debate case-per-case. 

An example of something of that nature you changed for the better is the way paths are built: using `os.sep` and `os.path.join` is of course the way to go for interoperability. I originally used `/` on purpose for improved readability / simplicity. Here, the benefit clearly outweighs the original intent.

What do you think of moving things like `Snek._find_files()`, `Snek.__get_nested_keys_from_filepath()` and `Snek.__update_attr_from_nested_keys()` to a separate `snekutils` file, as to reduce the amount of methods we have in `Snek`?

I added a few comments here and there on the PR as well.

Thanks again,

File: snek/snek.py (hunk 8 above)
Comment (timoguic) [reply snek/snek.py:451]: A bit of OCD on my side here. The file pointer is cleaned up at the end of the execution, so it is potentially left hanging for no good reason. If we have a very large number of files to process / write, that could become a problem. It should actually show up as a warning when running the test suite - I think that's what compelled me to change it :)

Comment (matteocargnelutti) [reply snek/snek.py:451]: It is good practice indeed. I _assume_ garbage collection catches it quickly after as we are in a loop context and the pointer is not referenced by anything, but better safe than debugging leaks ;) .

**Comment by timoguic** [comment]
Thanks for the feedback :)

I am not sure the `os` changes were actually required. I think `/` kind of works on Windows anyway, but I cannot tell for sure (or if that is actually a target platform for Snek).

I had the KISS in mind when I was working on the changes, and I hope I did not go too far. Some features have been abstracted (I believe it is easier to add support for more file formats with this) - but it comes at the cost of some "complexity". I like reasonably small modules, and I had the idea about splitting some of the code into a helper / lib package... But maybe you wanted a "one-file-to-rule-them-all" kind of approach? If we were to go the `snekutils` route, how would you see it? A class with static methods, plain functions?

A note about file naming: we have `snekconfig.py`, `snek.py`, potentially `snekutils.py`. Don't you want to keep it small and simple, `config.py`, `snek.py`, `utils.py`? We might end up with `snekbuilder.py`, `snekapi.py` and what not. That's a lot of sneks 🐍. :)

**Comment by matteocargnelutti** [comment]
> I am not sure the `os` changes were actually required. I think `/` kind of works on Windows anyway, but I cannot tell for sure 

That's what I assumed originally 😄 and I think it is true.

> (or if that is actually a target platform for Snek).
 
I wanted to target UNIX-y systems only, since Windows devs now use WSL extensively.

Also, all of the [snektools](https://github.com/matteocargnelutti/snektools) are super basic bash scripts: I don't think it would be worth adding Powershell versions of them.

> I had the KISS in mind when I was working on the changes, and I hope I did not go too far. Some features have been abstracted (I believe it is easier to add support for more file formats with this) - but it comes at the cost of some "complexity".

Nope, approach much appreciated 👍 .

> I like reasonably small modules, and I had the idea about splitting some of the code into a helper / lib package... But maybe you wanted a "one-file-to-rule-them-all" kind of approach? If we were to go the `snekutils` route, how would you see it? A class with static methods, plain functions?

Since it would be a collection of independent tools, I would see it as a simple file containing functions. We could then simply take advantage of Python's namespace features to keep them in context:
```
import utils

utils.super_tool(x)
```

That said. There could be value in having a class with static methods here to force scoping and to keep the pattern of _"1 file = 1 class"_. It really depends on the naming conventions we choose.

> A note about file naming: we have `snekconfig.py`, `snek.py`, potentially `snekutils.py`. Don't you want to keep it small and simple, `config.py`, `snek.py`, `utils.py`? We might end up with `snekbuilder.py`, `snekapi.py` and what not. That's a lot of sneks 🐍. :)

Ah ah ! Good point. 
So I started with `snek.py` and `snekconfig.py` to match the pattern _"1 file = 1 class"_, with the file named after the class, and of course assuming that the framework will remain fairly minimal by nature.

I do like having a prefix on most things as to avoid collisions with generic terms, but am open to the idea of having shorter filenames nevertheless.

Were you thinking of, for example, having `config.py` containing `SnekConfig`, or would you consider renaming the class `Config` as well?

Thanks again :) cheers.

**Comment by timoguic** [comment]
> I wanted to target UNIX-y systems only, since Windows devs now use WSL extensively.

Do you want me to revert these changes? Now that it's there... :)

> That said. There could be value in having a class with static methods here to force scoping and to keep the pattern of _"1 file = 1 class"_. It really depends on the naming conventions we choose.

Yup. I like namespacing - and we can always use `__all__`.

> Were you thinking of, for example, having `config.py` containing `SnekConfig`, or would you consider renaming the class `Config` as well?

Thanks for pointing out my lack of consistency 😞 I would go with `class SnekConfig` in `config.py`...

**Comment by matteocargnelutti** [comment]
> Do you want me to revert these changes? Now that it's there... :)

Absolutely not, it's good that way 😎 . Thank you !

> Yup. I like namespacing - and we can always use __all__.

Let's go for a _"static"_ class then 👍 .

> Thanks for pointing out my lack of consistency 😞 I would go with class SnekConfig in config.py...

That works for me 👍 .

Thanks again for your help !

**Comment by matteocargnelutti** [comment]
@timoguic Finally got a bit of time to have a look :D all good for me so far. Thank you very much for you help 👍

**Pull request closed without merging**